### DISCUSSION TITLE
Problem with language stats even when using .gitattributes

### DISCUSSION THREAD

**Discussion by recleun** [github-linguist/linguist#6211] [Classification] [answered]
## My problem

Hello, I currently have a repository and I'm having trouble with its language stats. The stats was originally 100% JavaScript, even though it should have been 100% TypeScript. I used the troubleshooting page on this repository to try and fix it, and then found out I can see which files are put in the statistics by clicking on the language of the repository and I did so, but another problem I found is that `index.ts` counts as a `.js` file (there were more `.ts` files but this is the only one that counted as `.js`), and also that `node_modules` folder is also put into the statistics. So I looked around and found out that I can use `.gitattributes` to fix these problems. I used `node_modules/** linguist-vendored` to exclude the `node_modules` folder, and then I used `*.ts linguist-language=TypeScript` to fix `index.ts` being counted as a `.js` file, and lastly I used `index.js linguist-generated=true` because it is basically `index.ts` but transpiled into JavaScript. I committed and pushed `.gitattributes` but then no changes were made to the statistics. Then I realized I could select the language I want from the code search/language menu, but I can only select one.

## What I need to know

1. Why did `.gitignore` have any effects on the language statistics.
2. How can I select multiple languages to be put into statistics.

## Extra notes:

`.gitignore` wasn't added when the repository was created, I manually added it.
I also have searched for a similar problem like this but didn't find anything useful.

### Edit:

Repository: https://github.com/recleun/wikisearch

**Comment by lildude** [comment, marked as answer]
As your repo stands, if you removed the `.gitattributes` file only two files will be considered by Linguist: the two `index.*` files. Everything else is already ignored by default, though keep in mind clicking the language actually takes you to Search which doesn't ignore the files like Linguist does else you'd never find them.

So the reason you only saw JavaScript initially and needed to implement an override for your TypeScript file is because of the very first [line in the file](https://github.com/recleun/wikisearch/blob/2ce2e45b3e8830342b3ac145d64e652a082e11d3/src/index.ts#L1): `#! /usr/bin/env node`.

This is a modeline indicating the interpreter of the file. `node` is not a valid TypeScript interpreter and is thus only associated with JavaScript in Linguist, however modelines take precedence as they're explicitly added by authors, so you effectively explicitly told Linguist to classify the file as JavaScript. 

With that said, I'm sure you can work out the answers I'm about to give to your questions…

 > Why did `.gitignore` have any effects on the language statistics.

Your `.gitattributes` was having an effect but you were however looking at the Search results which don't honour things like vendoring as then you wouldn't find your files. Also keep in mind, Search can take some time to reindex your repo.

> How can I select multiple languages to be put into statistics.

Remove the modeline or implement the override you did to force Linguist to ignore the modeline as I have done in my fork of your repo [here](https://github.com/lildude/wikisearch/tree/master) - I'll delete this fork when we're done here.

**Reply by recleun** [reply]
Thank you for the answer, I'm almost where I want to be. The problem is that whenever I run the index.js file, it gets executed by windows script host. Is there a way to prevent that from happening without using that modeline? Or is there a valid modeline for TypeScript?

**Reply by lildude** [reply]
You can leave it in and keep the functionality you're used to, but use an override to tell Linguist what to do. If you really want a valid TypeScript and JavaScript interpreter then [Deno](https://deno.land/) might do the trick (I'm not that experienced with JS/TS so I could be wrong) but with the caveat that Linguist only associates Deno with TypeScript so unless you're ignoring or overriding the generated `index.js` file, you'll just flip the behaviour.

**Reply by recleun** [reply]
Alright, that clears everything up for me. Thanks for the help.
